### ISSUE TITLE
MDB_KEYEXIST: Key/data pair already exists.

### ISSUE THREAD

**Issue by Ardnozz** [meilisearch/meilisearch#4300]
**Describe the bug**
I am experiencing a bug when I cannot post any request to meilisearch database because of the addition, replacement, and deletion of the only one lasting document in an index. The log of the middleware inside meilisearch is still registering them, but it is marking them as failed with unreasonable error.

**To Reproduce**
1. Create an empty index and set the primary key for "uuid"
2. Add one document to the index
3. Replace the document with another value
4. Delete the document (so there is no document inside the index)
5. Add a new document and instead of its creation you will see a failed task (even though the uuid is different, this should not be the cause)

**Expected behavior**
The database should add new documents without any errors and failed tasks.

**Screenshots**
Failed task
![image](https://github.com/meilisearch/meilisearch/assets/94400798/38078799-9001-4f83-a046-daee9e8200e5)

**Meilisearch version:**
v1.5.0

**Additional context**
I am using the docker version of meilisearch built on WSL and for posting the request to the database I use an official meilisearch-js API.

**Comment by dureuill** [comment]
Hello @Ardnozz 👋 

Thank you very for the very actionable report ✨ 

It looks like you encountered an error with the engine, I'm sorry that this is happening 😞 

Unfortunately, even by following your detailed steps, I cannot reproduce the issue you're facing, be it on v1.5.0 or on v1.6.0-rc.4.

Here is the list of `curl` commands I executed:

```sh
# 1. Create an empty index and set the primary key for "uuid"
curl http://localhost:7700/indexes -H 'Content-Type: application/json' --data-binary '{"uid": "lecturers", "primaryKey": "uuid"}'
# 2. Add one document to the index
curl http://localhost:7700/indexes/lecturers/documents -H 'Content-Type: application/json' --data-binary '[{"uuid": "510fde4e-a07a-4676-8b07-680f32f3ecd7", "title": "initial document"}]'
# 3. Replace the document with another value
curl http://localhost:7700/indexes/lecturers/documents -H 'Content-Type: application/json' --data-binary '[{"uuid": "510fde4e-a07a-4676-8b07-680f32f3ecd7", "title": "replaced document"}]'
# 4. Delete the document (so there is no document inside the index)
curl -X DELETE http://localhost:7700/indexes/lecturers/documents/510fde4e-a07a-4676-8b07-680f32f3ecd7
# 5. Add a new document
curl http://localhost:7700/indexes/lecturers/documents -H 'Content-Type: application/json' --data-binary '[{"uuid": "d50708bb-2893-474f-90ee-58d3dba39900", "title": "after delete"}]'
```

Could you perhaps check that you do reproduce the issue using the above reproducer? If not, is there any step I misinterpreted? Can you modify the script above to trigger the issue?

I really need to be able to reproduce so that I can see what is happening and hopefully find a fix 🙁 

Thank you for your cooperation, and again your great report ❤️

**Comment by Ardnozz** [comment]
Hello @dureuill 

I've tried to reproduce the bug with your curl commands and everything works perfectly. So I checked my system again and I found out that the problem is in `updateFilterableAttributes` action. Here is a list of `curl` commands that worked for me. The only difference is the API key, but I believe that it cannot do that much different.
```shell
# 1. Create an empty index and set the primary key to "uuid"
curl --location 'http://localhost:7700/indexes' --header 'Content-Type: application/json' --header 'Authorization: Bearer OURSUPERSECRETKEY' --data '{"uid": "lecturers", "primaryKey": "uuid"}'
#2. Update filterable attributes with "uuid"
curl --location --request PUT 'http://localhost:7700/indexes/lecturers/settings/filterable-attributes' --header 'Content-Type: application/json' --header 'Authorization: Bearer OURSUPERSECRETKEY' --data '["uuid"]'
#3. Add one document to the index
curl --location 'http://localhost:7700/indexes/lecturers/documents' --header 'Content-Type: application/json' --header 'Authorization: Bearer OURSUPERSECRETKEY' --data '[{"uuid": "2e7e7f4c-9402-4abf-9f5d-eb9157ee79a5", "title": "Initial title"}]'
#4. Replace the document in the index
curl --location 'http://localhost:7700/indexes/lecturers/documents' --header 'Content-Type: application/json' --header 'Authorization: Bearer OURSUPERSECRETKEY' --data '[{"uuid": "2e7e7f4c-9402-4abf-9f5d-eb9157ee79a5", "title": "Next title"}]'
#5. Delete the only document
curl --location --request DELETE 'http://localhost:7700/indexes/lecturers/documents/2e7e7f4c-9402-4abf-9f5d-eb9157ee79a5' --header 'Authorization: Bearer OURSUPERSECRETKEY'
#6. Try adding new document with completly new "uuid" and "title" property.
curl --location 'http://localhost:7700/indexes/lecturers/documents' --header 'Content-Type: application/json' --header 'Authorization: Bearer OURSUPERSECRETKEY' --data '[{"uuid": "510fde4e-a07a-4676-8b07-680f32f3ecd7", "title": "Initial title2"}]'
```
Hope this will work (or don't work 😅) as expected. _(The curls are autogenerated by Postman)_
Ardno

**Comment by dureuill** [comment]
Hello @Ardnozz 

Unfortunately, I'm still unable to reproduce the issue 🥺

I ran the reproducer you provided as a bash script, but all my tasks still succeeded.
I retried multiple times on fresh DBs by interspacing a task on an unrelated index to check if it could be autobatching shenanigans, but the behavior is stable (success) regardless of the task batching behavior.

Can you confirm that running the reproducer from your message on a fresh Meilisearch DB results in exposing the issue?

If so, maybe something with your setup is causing the issue 🤔 

The only peculiar thing I can see is that you're running WSL (also, you're using docker and I'm running a local v1.5 version on Mac).

1. Are you running WSL 1 or WSL 2? (if you don't know, that's probably WSL2)
2. if running WSL2, which distribution are you using? 
3. Do you have the possibility and the time to try the reproducer in a different environment?

If not, I'll try to reproduce your setup on my side.

Thanks a lot for your reactivity and help in getting to the bottom of this 💯

**Comment by Ardnozz** [comment]
You're supposing right. I'm using WSL with Docker.
```yml
# wsl --version
Version WSL: 1.2.5.0
Version core: 5.15.90.1
Version WSLg: 1.0.51
Version MSRDC: 1.2.3770
Version Direct3D: 1.608.2-61064218
Version DXCore: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Version Windows: 10.0.19045.3803
```
```yml
# wsl --status
Default distro: docker-desktop-data
Default version: 2
```
We've tried the same version on Linux and it works as expected. Maybe if you wouldn't be able to reproduce it, I can switch to a new version and check if that works.

**Comment by Ardnozz** [comment]
I've tried upgrading to the latest version `v1.6.0-rc.6` and built it from `Dockerfile` for me and downloaded the binary located in `/bin/meilisearch` so it can be as a part of another Docker image. The problem persists.

**Comment by balach** [comment]
Unfortunately, I ran into the same issue just now - using Meilisearch Cloud with 1.6.0, with using firestore with the official firebase extension.. 

Important: We did indeed delete all documents in our index and then repopulated our databse because we were going live today. And since then the index does not update and throws the error.

The full error:
{
  "indexUid": "propertiesIndex",
  "status": "failed",
  "type": "documentAdditionOrUpdate",
  "uid": 3934,
  "details": {
    "receivedDocuments": 1,
    "indexedDocuments": 0
  },
  "canceledBy": null,
  "error": {
    "message": "internal: MDB_KEYEXIST: Key/data pair already exists.",
    "code": "internal",
    "type": "internal",
    "link": "https://docs.meilisearch.com/errors#internal"
  },
  "duration": "PT0.080720649S",
  "startedAt": "2024-01-18T19:42:54.776Z",
  "enqueuedAt": "2024-01-18T19:42:54.773Z",
  "finishedAt": "2024-01-18T19:42:54.857Z"
}

**Comment by queer** [comment]
I ran into this today as well, using the `1.6.0-1` deb package. Exact same error that @balach reported. Deleting the entire index and recreating it fixed the problem.

**Comment by mech** [comment]
I also encounter this issues today with "MDB_KEYEXIST: key/data pair already exists"

I am using macOS Monterey, Docker Desktop v4.27.1 with Meilisearch v1.6.1

Doing a clear_index! and reindex! again fix the issues. But I am now afraid that it will suddenly appear in production.

**Comment by Kerollmops** [comment]
Hey @mech 👋

Thank you for the report! However, can you help us reproduce the issue? Can you share your settings? The documents you were sending and the way you were sending them (batch size)?

Have a nice day 🌻

**Comment by queer** [comment]
I ran into this again yesterday (still on `1.6.0`) and got some more details that may help:

- Documents are ingested in batches of 500~2500 at a time, ~540k total documents.
- Documents are often replacing existing documents, rather than inserting new documents.
- All documents have the exact same fields.
- When checking stats (`GET /stats`), I see:
  ```js
  {
    "databaseSize": 6392389632,
    "lastUpdate": "2024-02-14T22:27:15.981578502Z",
    "indexes": {
      "products": {
        "numberOfDocuments": 0,
        "isIndexing": false,
        "fieldDistribution": {
          "_geo": 540925,
          "brand": 540925,
          // Continues for all 19 fields that present
        }
      }
    }
  }
  ```
  Note that `numberOfDocuments` is 0, but `fieldDistribution` shows the expected numbers.
- This seems to only happen when updating documents, never on the initial insert of documents into the index.

**Comment by opt76562a** [comment]
Thank you for that info. So is this a good or bad think for done-for-you system? For online email affiliate marketing?

**Comment by dureuill** [comment]
Hello all of you :wave:

We recently released [v1.7.0-rc.1](https://github.com/meilisearch/meilisearch/releases/tag/v1.7.0-rc.1), a pre-release version that features changes susceptible of fixing this issue.

Since we are not reproducing ourselves, could some of you who experience the issue take the time to test the new version and report if the issue still reproduces? 🙏 

We would love that :heart:

⚠️ **Please do not use a pre-release version in a production environment, this is for testing purposes only.** The final v1.7.0 version is stated for release on March, 11th

**Comment by nlgtuankiet** [comment]
Hi @Ardnozz did you try out the v1.7.0-rc.1 version to see if this issue is fixed?

**Comment by queer** [comment]
This issue stopped reproducing on v1.6.2 for me at some point; unfortunately I never managed to pinpoint the specific problematic document.

**Comment by Kerollmops** [comment]
Hey @queer, if you are unable to reproduce on v1.6.2, it's because we fixed it in this version by disabling some parts of the engine. However, we perfectly fixed it in v1.7, and indexing is faster on this one too.
